File: mcp/src/minecraft/femtocraft/research/containers/ContainerResearchConsole.java
/*******************************************************************************
 * Copyright (C) 2013  Christopher Harris (Itszuvalex)
 * Itszuvalex@gmail.com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 ******************************************************************************/

package femtocraft.research.containers;

import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;
import femtocraft.common.gui.OutputSlot;
import femtocraft.research.tiles.TileEntityResearchConsole;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.entity.player.InventoryPlayer;
import net.minecraft.inventory.Container;
import net.minecraft.inventory.ICrafting;
import net.minecraft.inventory.Slot;
import net.minecraft.item.ItemStack;
import net.minecraft.item.crafting.FurnaceRecipes;

public class ContainerResearchConsole extends Container {
    private static final int progressID = 0;
    private static final int progressMaxID = 1;
    private final TileEntityResearchConsole console;
    private int lastProgress = 0;
    private int lastProgressMax = 0;

    public ContainerResearchConsole(InventoryPlayer par1InventoryPlayer,
                                    TileEntityResearchConsole console) {
        this.console = console;

        this.addSlotToContainer(new OutputSlot(console, 9, 147, 60));

        for (int i = 0; i < 9; ++i) {
            this.addSlotToContainer(new Slot(console, i, 8 + 18 * (i % 3),
                                             16 + 18 * (i / 3)));
        }

        int i;

        // Bind player inventory
        for (i = 0; i < 3; ++i) {
            for (int j = 0; j < 9; ++j) {
                this.addSlotToContainer(new Slot(par1InventoryPlayer, j + i * 9
                        + 9, 8 + j * 18, 84 + i * 18));
            }
        }
        // Bind player actionbar
        for (i = 0; i < 9; ++i) {
            this.addSlotToContainer(new Slot(par1InventoryPlayer, i,
                                             8 + i * 18, 142));
        }
    }

    @Override
    public boolean canInteractWith(EntityPlayer entityplayer) {
        return console.isUseableByPlayer(entityplayer);
    }

    @Override
    @SideOnly(Side.CLIENT)
    public void updateProgressBar(int par1, int par2) {
        super.updateProgressBar(par1, par2);

        switch (par1) {
            case progressID:
                console.setResearchProgress(par2);
                break;
            case progressMaxID:
                console.setResearchMax(par2);
                break;
        }
    }

    @Override
    public void detectAndSendChanges() {
        super.detectAndSendChanges();

        for (Object crafter : this.crafters) {
            ICrafting icrafting = (ICrafting) crafter;

            if (this.lastProgress != this.console.getResearchProgress()) {
                icrafting.sendProgressBarUpdate(this, progressID,
                                                this.console.getResearchProgress());
            }
            if (this.lastProgressMax != this.console.getResearchMax()) {
                icrafting.sendProgressBarUpdate(this, progressMaxID,
                                                this.console.getResearchMax());
            }
        }

        this.lastProgress = this.console.getResearchProgress();
        this.lastProgressMax = this.console.getResearchMax();
    }

    @Override
    public void addCraftingToCrafters(ICrafting par1iCrafting) {
        super.addCraftingToCrafters(par1iCrafting);
        par1iCrafting.sendProgressBarUpdate(this, progressID,
                                            console.getResearchProgress());
        par1iCrafting.sendProgressBarUpdate(this, progressMaxID,
                                            console.getResearchMax());
    }

    /**
     * Called when a player shift-clicks on a slot. You must override this or
     * you will crash when someone does that.
     */
    @Override
    public ItemStack transferStackInSlot(EntityPlayer par1EntityPlayer, int par2) {
        ItemStack itemstack = null;
        Slot slot = (Slot) this.inventorySlots.get(par2);

        if (slot != null && slot.getHasStack()) {
            ItemStack itemstack1 = slot.getStack();
            itemstack = itemstack1.copy();

            if (par2 == 1) {
                if (!this.mergeItemStack(itemstack1, 2, 38, true)) {
                    return null;
                }

                slot.onSlotChange(itemstack1, itemstack);
            }
            else if (par2 != 0) {
                if (FurnaceRecipes.smelting().getSmeltingResult(itemstack1) != null) {
                    if (!this.mergeItemStack(itemstack1, 0, 1, false)) {
                        return null;
                    }
                }
                else if (par2 >= 2 && par2 < 29) {
                    if (!this.mergeItemStack(itemstack1, 29, 38, false)) {
                        return null;
                    }
                }
                else if (par2 >= 29 && par2 < 38
                        && !this.mergeItemStack(itemstack1, 2, 29, false)) {
                    return null;
                }
            }
            else if (!this.mergeItemStack(itemstack1, 2, 38, false)) {
                return null;
            }

            if (itemstack1.stackSize == 0) {
                slot.putStack(null);
            }
            else {
                slot.onSlotChanged();
            }

            if (itemstack1.stackSize == itemstack.stackSize) {
                return null;
            }

            slot.onPickupFromSlot(par1EntityPlayer, itemstack1);
        }

        return itemstack;
    }
}

File: src/main/java/com/itszuvalex/femtocraft/research/containers/ContainerResearchConsole.scala
/*
 * ******************************************************************************
 *  * Copyright (C) 2013  Christopher Harris (Itszuvalex)
 *  * Itszuvalex@gmail.com
 *  *
 *  * This program is free software; you can redistribute it and/or
 *  * modify it under the terms of the GNU General Public License
 *  * as published by the Free Software Foundation; either version 2
 *  * of the License, or (at your option) any later version.
 *  *
 *  * This program is distributed in the hope that it will be useful,
 *  * but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  * GNU General Public License for more details.
 *  *
 *  * You should have received a copy of the GNU General Public License
 *  * along with this program; if not, write to the Free Software
 *  * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *  *****************************************************************************
 */
package com.itszuvalex.femtocraft.research.containers

import com.itszuvalex.femtocraft.common.gui.OutputSlot
import com.itszuvalex.femtocraft.core.container.ContainerBase
import com.itszuvalex.femtocraft.research.containers.ContainerResearchConsole._
import com.itszuvalex.femtocraft.research.tiles.TileEntityResearchConsole
import cpw.mods.fml.relauncher.{Side, SideOnly}
import net.minecraft.entity.player.{EntityPlayer, InventoryPlayer}
import net.minecraft.inventory.{ICrafting, Slot}
import net.minecraft.item.ItemStack
import net.minecraft.item.crafting.FurnaceRecipes

import scala.collection.JavaConversions._

object ContainerResearchConsole {
  private val progressID    = 0
  private val progressMaxID = 1
}

class ContainerResearchConsole(par1InventoryPlayer: InventoryPlayer, private val console: TileEntityResearchConsole) extends ContainerBase {
  private var lastProgress    = 0
  private var lastProgressMax = 0

  addSlotToContainer(new OutputSlot(console, 9, 147, 60))
  for (i <- 0 until 9) {
    addSlotToContainer(new Slot(console, i, 8 + 18 * (i % 3), 16 + 18 * (i / 3)))
  }

  for (i <- 0 until 3) {
    for (j <- 0 until 9) {
      addSlotToContainer(new Slot(par1InventoryPlayer, j + i * 9 + 9, 8 + j * 18, 84 + i * 18))
    }
  }
  for (i <- 0 until 9) {
    addSlotToContainer(new Slot(par1InventoryPlayer, i, 8 + i * 18, 142))
  }


  override def addCraftingToCrafters(par1iCrafting: ICrafting) {
    super.addCraftingToCrafters(par1iCrafting)
    sendUpdateToCrafter(this, par1iCrafting, progressID, console.getResearchProgress)
    sendUpdateToCrafter(this, par1iCrafting, progressMaxID, console.getResearchMax)
  }

  override def detectAndSendChanges() {
    super.detectAndSendChanges()
    crafters.foreach { case icrafting: ICrafting =>
      if (lastProgress != console.getResearchProgress) {
        sendUpdateToCrafter(this, icrafting, progressID, console.getResearchProgress)
      }
      if (lastProgressMax != console.getResearchMax) {
        sendUpdateToCrafter(this, icrafting, progressMaxID, console.getResearchMax)
      }
                     }
    lastProgress = console.getResearchProgress
    lastProgressMax = console.getResearchMax
  }

  /**
   * Called when a player shift-clicks on a slot. You must override this or you will crash when someone does that.
   */
  override def transferStackInSlot(par1EntityPlayer: EntityPlayer, par2: Int): ItemStack = {
    var itemstack: ItemStack = null
    val slot = inventorySlots.get(par2).asInstanceOf[Slot]
    if (slot != null && slot.getHasStack) {
      val itemstack1 = slot.getStack
      itemstack = itemstack1.copy
      if (par2 == 1) {
        if (!mergeItemStack(itemstack1, 2, 38, true)) {
          return null
        }
        slot.onSlotChange(itemstack1, itemstack)
      }
      else if (par2 != 0) {
        if (FurnaceRecipes.smelting.getSmeltingResult(itemstack1) != null) {
          if (!mergeItemStack(itemstack1, 0, 1, false)) {
            return null
          }
        }
        else if (par2 >= 2 && par2 < 29) {
          if (!mergeItemStack(itemstack1, 29, 38, false)) {
            return null
          }
        }
        else if (par2 >= 29 && par2 < 38 && !mergeItemStack(itemstack1, 2, 29, false)) {
          return null
        }
      }
      else if (!mergeItemStack(itemstack1, 2, 38, false)) {
        return null
      }
      if (itemstack1.stackSize == 0) {
        slot.putStack(null)
      }
      else {
        slot.onSlotChanged()
      }
      if (itemstack1.stackSize == itemstack.stackSize) {
        return null
      }
      slot.onPickupFromSlot(par1EntityPlayer, itemstack1)
    }
    itemstack
  }

  @SideOnly(Side.CLIENT) override def updateProgressBar(par1: Int, par2: Int) {
    super.updateProgressBar(par1, par2)
    par1 match {
      case `progressID` => console.setResearchProgress(par2)
      case `progressMaxID` => console.setResearchMax(par2)
    }
  }

  def canInteractWith(entityplayer: EntityPlayer) = console.isUseableByPlayer(entityplayer)
}
